llvm/lib/Target/AIE/AIELegalizerHelper.cpp

```diff
@@ -182,9 +185,32 @@ bool AIELegalizerHelper::legalizeG_BUILD_VECTOR(LegalizerHelper &Helper,
   //    https://github.com/llvm/llvm-project/pull/84538
   if (DstVecSize == 256) {
     const Register UnusedSubReg = MRI.createGenericVirtualRegister(DstVecTy);
-    MIRBuilder.buildUnmerge({DstReg, UnusedSubReg}, Src);
+    MIRBuilder.buildUnmerge({UnusedSubReg, DstReg}, Src);
   } else if (DstVecSize == 128) {
-    MIRBuilder.buildInstr(II->getGenericUnpadVectorOpcode(), {DstReg}, {Src});
+    const LLT V16S32 = LLT::fixed_vector(16, 32);
+    Register Vec512Reg = MRI.createGenericVirtualRegister(V16S32);
+
+    Register Zero = MIRBuilder.buildConstant(LLT::scalar(32), 0).getReg(0);
+    Register ShiftConstant =
+        MIRBuilder.buildConstant(LLT::scalar(32), 48).getReg(0);
+
+    Register NewSrc1 =
+        EltSize == 32 ? Src : MIRBuilder.buildBitcast(V16S32, Src).getReg(0);
+
+    // Shift the result to the lower 128 bits of a 512-bit vector, as the
+    // result is currently stored in the upper 128 bits of the vector.
+    MIRBuilder
+        .buildIntrinsic(Intrinsic::aie2_vshift_I512_I512, Vec512Reg, false,
```

Comment:
I don't think we should bring an AIE2-specific intrinsic here, this code should stay version agnostic I believe